Fix kernel memory limit

hqhq opened the pull request:
Seems it never works, don't how that happens. :(

Signed-off-by: Qiang Huang h.huangqiang@huawei.com

runcom commented:
@hqhq could you please add an integration test which start a container with `--kernel-memory=VALUE` and `cat /sys/fs/cgroup/memory/memory.kmem.tcp.limit_in_bytes` and makes sure `VALUE

hqhq commented:
@runcom Sure, will do tomorrow, thanks.

hqhq committed:
Fix kernel memory limit

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq commented:
@runcom Updated, and I also updated other cgroup integration tests.

runcom commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -134,25 +134,24 @@ func (s *DockerSuite) TestRunAttachDetach(c *check.C) {
-	c.Assert(out, checker.Equals, "test", check.Commentf("container should've printed 'test'"))
+	file := "/sys/fs/cgroup/cpu/cpu.cfs_quota_us"
+	out, _ := dockerCmd(c, "run", "--cpu-quota", "8000", "--name", "test", "busybox", "cat", file)
+	c.Assert(out, checker.Contains, "8000")
```
Shouldn't make much difference but maybe use `Equals` instead of `Contains`?

hqhq replied:
In case `-m`, I got some warnings about memory swap is not supported in `out`, which is intended, so I use `Contains` instead, maybe we can use `Equals` for other cases that won't cause any warnings because of the different kernel support.

runcom replied:
Mmm right, I think the warning is printed to the stderr so if you only get the stdout from the command you can use Equals, but nvm it's just a nit

runcom commented:
Just a nit, LGTM otherwise

hqhq commented:
@runcom Updated, use `Contains` only for `-m` case, change others to `Equals`.

thaJeztah commented:
Added this to the 1.9.2 milestone in case we release a 1.9.2 release (undecided yet)

runcom commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -238,13 +265,17 @@ func (s *DockerSuite) TestRunOOMExitCode(c *check.C) {
+	file := "/sys/fs/cgroup/memory/memory.limit_in_bytes"
+	out, _ := dockerCmd(c, "run", "-m", "32M", "--name", "test", "busybox", "cat", file)
+	// Use `Contains` here because we could got warnings about memory swap not supported in `out`
+	c.Assert(out, checker.Contains, "33554432")
```
```
stdout, _, err := dockerCmdWithStdoutStderr(.....)
```
I think with `stdout` you can use Equals here (I can't try right now)

runcom replied:
But really is just a nit ahah

runcom replied:
Tests are failing because `out` needs `TrimSpace`

hqhq committed:
Update cgroup integration tests

Cgroup integtaion tests should cover:
- docker can run sucessfully with these options
- these cgroup options are set to HostConfig as expected
- these cgroup options are really set to cgroup files as expected
- other cases (wrong value, combinations etc..)

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq commented:
@runcom Updated, thanks :)

runcom commented:
re LGTM

estesp commented:
LGTM
